### PR TITLE
[Proof of Concept] speedup checkpoint running with multiple threads

### PR DESCRIPTION
**Pull request by jdimatteo**
This branch is an experiment to collect data as presented in this PR. So far no attempt was made to ensure thread safety and this PR is not currently in a state to be merged. The benchmarks analyzed here are added in https://github.com/great-expectations/great_expectations/pull/3280.

### pytest-benchmark wall clock times

The below pytest-benchmark wall clock times shows the following:
* using 100 threads speeds up execution ~20x when not rendering HTML data docs (`624.1411/31.1652=20.03`)
* using 100 threads speeds up execution ~7x when including rendering HTML data docs (`665.2896/93.9458=7.08`)

```
(venv) jdimatteo@hcgr:~/dev/great_expectations_fork/tests/performance/results$ py.test-benchmark compare --name long --columns=min *html*

----------------------- benchmark: 11 tests -----------------------------------------------------------------------
Name (time in s)                                                                                      Min          
-------------------------------------------------------------------------------------------------------------------
test_bikeshare_trips_benchmark[100] (no_html_run_1_100_threads_unlimitted_pool_cloudtop)          31.1652 (1.0)    
test_bikeshare_trips_benchmark[100] (no_html_run_2_100_threads_unlimitted_pool_cloudtop)          31.5370 (1.01)   
test_bikeshare_trips_benchmark[100] (with_html_run_1_100_threads_pool_size_2_cloudtop)            93.9458 (3.01)   
test_bikeshare_trips_benchmark[100] (with_html_run_2_100_threads_unlimitted_pool_cloudtop)        94.1783 (3.02)   
test_bikeshare_trips_benchmark[100] (with_html_run_2_100_threads_pool_size_2_cloudtop)            94.2633 (3.02)   
test_bikeshare_trips_benchmark[100] (with_html_run_1_100_threads_unlimitted_pool_cloudtop)        96.6022 (3.10)   
test_bikeshare_trips_benchmark[100] (no_html_50_threads_unlimitted_pool_cloudtop)                165.6627 (5.32)   
test_bikeshare_trips_benchmark[100-False] (baseline_single_threaded_cloudtop_no_html_run2)       624.1411 (20.03)  
test_bikeshare_trips_benchmark[100-False] (baseline_single_threaded_cloudtop_no_html)            659.4806 (21.16)  
test_bikeshare_trips_benchmark[100-True] (baseline_single_threaded_cloudtop_with_html_run_2)     665.2896 (21.35)  
test_bikeshare_trips_benchmark[100-True] (baseline_single_threaded_cloudtop_with_html)           668.8410 (21.46)  
-------------------------------------------------------------------------------------------------------------------
```

#### Single Thread

Based on the below command running the benchmark with a single thread and output:
* CPU time accounts for only 5% of wall time (`32.831/634.948~=.05`).
    * The total wall time is 634.948 seconds (`10*60+34.948=634.948`).
    * The total CPU time is 32.831 seconds (`27.499+5.332=32.831`).
* The pytest-benchmark wall clock time accounts for 98% of the wall clock time reported by the Unix `time` command (`624.1411/634.948~=0.98`).
    * The remaining 2% of the time is assumed to be setup, the assert statements, starting python, etc.

```
(venv) jdimatteo@hcgr:~/dev/great_expectations_fork$ time pytest tests/performance/test_bigquery_benchmarks.py   --bigquery --performance-tests --benchmark-json=tests/performance/results/baseline_single_threaded_cloudtop_no_html_run2.json -k test_bikeshare_trips_benchmark[100-False]  --no-spark --no-postgresql -rP -vv
```

Above command ended with the following output:

```
----------------------------------------------------------- benchmark: 1 tests -----------------------------------------------------------
Name (time in s)                                   Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------
test_bikeshare_trips_benchmark[100-False]     624.1411  624.1411  624.1411  0.0000  624.1411  0.0000       0;0  0.0016       1           1
------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
================================================================================================================================= 1 passed, 11 deselected, 202 warnings in 629.54s (0:10:29) =================================================================================================================================

real    10m34.948s
user    0m27.499s
sys     0m5.332s
```

#### Multiple Threads

When using 50 threads, wall clock time decreased while CPU time increases based on the below command and output:
* Wall time decreased to 178.339 seconds (`2*60+58.339=178.339`).
* CPU time increased to 170.918 seconds (`2*60+19.077+31.841=170.918`).
* CPU time accounts for about 95% of wall time (`170.918/178.339~=.95`).

```
(venv) jdimatteo@hcgr:~/dev/great_expectations_fork$ time pytest tests/performance/test_bigquery_benchmarks.py   --bigquery --performance-tests --benchmark-json=tests/performance/results/no_html_50_threads_unlimitted_pool_cloudtop.json -k test_bikeshare_trips_benchmark[100]  --no-spark --no-postgresql -rP -vv
```

```
-------------------------------------------------------- benchmark: 1 tests --------------------------------------------------------
Name (time in s)                             Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------
test_bikeshare_trips_benchmark[100]     165.6627  165.6627  165.6627  0.0000  165.6627  0.0000       0;0  0.0060       1           1
------------------------------------------------------------------------------------------------------------------------------------
Calculating Metrics: 100%|██████████| 25/25 [00:14<00:00,  1.65it/s]
Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
========================================================================================= 1 passed, 6 deselected, 229 warnings in 172.43s (0:02:52) ==========================================================================================
    
real    2m58.339s
user    2m19.077s
sys     0m31.841s
```

## cProfile Results

[cProfile](https://docs.python.org/3/library/profile.html) results below measure wall time and shows most time being spent waiting for database execution to complete.

These profile results were from the baseline single threaded performance for a single table with the following command at https://github.com/great-expectations/great_expectations/pull/3280/commits/cfe290aa39d712d455b1500799ea7c7218ce7fb5:

```
GE_PROFILE_FILE_PATH=tests/performance/results/cprofile_`date "+%H%M"`_`git rev-parse --verify HEAD` pytest tests/performance/test_bigquery_benchmarks.py --no-spark --no-postgresql -rP -vv --bigquery --performance-tests -k "test_bikeshare_trips_benchmark[1-True]"
```

The below call graph visualization shown by [PyCharm](https://www.jetbrains.com/help/pycharm/profiler.html#view-graph) shows:
* At the bottom, [Checkpoint.run](https://github.com/great-expectations/great_expectations/blob/8112e73826a92c2e28f1c9d6e6d94811835c22f9/great_expectations/checkpoint/checkpoint.py#L211) takes 100% of the wall time, ~11.2 seconds
* Second from the bottom, [ActionListValidationOperator.run](https://github.com/great-expectations/great_expectations/blob/8112e73826a92c2e28f1c9d6e6d94811835c22f9/great_expectations/validation_operators/validation_operators.py#L276) takes 99.8% of the wall time (~11.2 seconds).
* Third row from the top, [ExecutionEngine.resolve_metrics](https://github.com/great-expectations/great_expectations/blob/2f4291ec7bc8fbe5f45e0b90ecadc15a522b4c21/great_expectations/execution_engine/execution_engine.py#L204) takes 92.6% of the wall time (~10.4 seconds).
* At the top, [SqlAlchemy Engine.execute](https://github.com/zzzeek/sqlalchemy/blob/rel_1_4_9/lib/sqlalchemy/engine/base.py#L3018) takes 69.7% of the wall time (~7.8 seconds).

cprofile_1123_higher_call_stack

Note: I uploaded [my full cProfile results](https://drive.google.com/file/d/1MweSLejs3UfTKf2jNlyRS243gulkFFLk/view?usp=sharing). It isn't very clear to me exactly what data is in these profile results, so I only shared this file with select individuals.

## Line By Line Results

[line_profiler](https://github.com/pyutils/line_profiler#line_profiler-and-kernprof) was run on [ActionListValidationOperator.run](https://github.com/great-expectations/great_expectations/blob/8112e73826a92c2e28f1c9d6e6d94811835c22f9/great_expectations/validation_operators/validation_operators.py#L276), which the cProfile results above showed accounting for 99.8 of the wall time. See https://github.com/great-expectations/great_expectations/pull/3280/commits/90d24741fa59003b0ae3b74ea551c7c702036935 for details on experiment setup and/or to reproduce results.

```
kernprof -o tests/performance/results/line_profiler_`date "+%H%M"`_`git rev-parse --verify HEAD`.lprof -l -v tests/performance/test_bigquery_benchmarks.py --no-spark --no-postgresql -rP -vv --bigquery --performance-tests -k "test_bikeshare_trips_benchmark[1-True]" 
```

Resulted in the following output, showing:
* 93.5% of the wall time is spent in the call to [batch.validate](https://github.com/great-expectations/great_expectations/blob/2f4291ec7bc8fbe5f45e0b90ecadc15a522b4c21/great_expectations/validator/validator.py#L1005) (and the cProfile results above show this is mostly spent resolving metrics through database execution).
* 6.5% of the wall time is spent in [ActionListValidationOperator._run_actions](https://github.com/great-expectations/great_expectations/blob/fb20bd56e08b1aa0d7ef159c38b3dd822be1ea1b/great_expectations/validation_operators/validation_operators.py#L350).

```
Wrote profile results to tests/performance/results/line_profiler_1215_90d24741fa59003b0ae3b74ea551c7c702036935.lprof
Timer unit: 1e-06 s

Total time: 13.1165 s
File: /home/jdimatteo/dev/great_expectations_fork_1/great_expectations/validation_operators/validation_operators.py
Function: run at line 276

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   276                                               @profile
   277                                               def run(
   278                                                   self,
   279                                                   assets_to_validate,
   280                                                   run_id=None,
   281                                                   evaluation_parameters=None,
   282                                                   run_name=None,
   283                                                   run_time=None,
   284                                                   result_format=None,
   285                                               ):
   286                                           
   287         3          4.0      1.3      0.0          assert not (run_id and run_name) and not (
   288         2          1.0      0.5      0.0              run_id and run_time
   289                                                   ), "Please provide either a run_id or run_name and/or run_time."
   290         1          1.0      1.0      0.0          if isinstance(run_id, str) and not run_name:
   291                                                       warnings.warn(
   292                                                           "String run_ids will be deprecated in the future. Please provide a run_id of type "
   293                                                           "RunIdentifier(run_name=None, run_time=None), or a dictionary containing run_name "
   294                                                           "and run_time (both optional). Instead of providing a run_id, you may also provide"
   295                                                           "run_name and run_time separately.",
   296                                                           DeprecationWarning,
   297                                                       )
   298                                                       try:
   299                                                           run_time = parse(run_id)
   300                                                       except (ValueError, TypeError):
   301                                                           pass
   302                                                       run_id = RunIdentifier(run_name=run_id, run_time=run_time)
   303         1          1.0      1.0      0.0          elif isinstance(run_id, dict):
   304                                                       run_id = RunIdentifier(**run_id)
   305         1          1.0      1.0      0.0          elif not isinstance(run_id, RunIdentifier):
   306                                                       run_id = RunIdentifier(run_name=run_name, run_time=run_time)
   307                                           
   308         1          1.0      1.0      0.0          run_results = {}
   309                                           
   310         2          2.0      1.0      0.0          for item in assets_to_validate:
   311         1          1.0      1.0      0.0              run_result_obj = {}
   312         1          4.0      4.0      0.0              batch = self._build_batch_from_item(item)
   313                                           
   314         1          8.0      8.0      0.0              if hasattr(batch, "active_batch_id"):
   315         1          2.0      2.0      0.0                  batch_identifier = batch.active_batch_id
   316                                                       else:
   317                                                           batch_identifier = batch.batch_id
   318                                           
   319         2          6.0      3.0      0.0              expectation_suite_identifier = ExpectationSuiteIdentifier(
   320         1          1.0      1.0      0.0                  expectation_suite_name=batch._expectation_suite.expectation_suite_name
   321                                                       )
   322         2         11.0      5.5      0.0              validation_result_id = ValidationResultIdentifier(
   323         1          0.0      0.0      0.0                  batch_identifier=batch_identifier,
   324         1          1.0      1.0      0.0                  expectation_suite_identifier=expectation_suite_identifier,
   325         1          0.0      0.0      0.0                  run_id=run_id,
   326                                                       )
   327         2   12263648.0 6131824.0     93.5              batch_validation_result = batch.validate(
   328         1          1.0      1.0      0.0                  run_id=run_id,
   329         1          1.0      1.0      0.0                  result_format=result_format if result_format else self.result_format,
   330         1          0.0      0.0      0.0                  evaluation_parameters=evaluation_parameters,
   331                                                       )
   332         1          2.0      2.0      0.0              run_result_obj["validation_result"] = batch_validation_result
   333         2     852732.0 426366.0      6.5              batch_actions_results = self._run_actions(
   334         1          0.0      0.0      0.0                  batch,
   335         1          1.0      1.0      0.0                  expectation_suite_identifier,
   336         1          1.0      1.0      0.0                  batch._expectation_suite,
   337         1          1.0      1.0      0.0                  batch_validation_result,
   338         1          1.0      1.0      0.0                  run_id,
   339                                                       )
   340                                           
   341         1          2.0      2.0      0.0              run_result_obj["actions_results"] = batch_actions_results
   342         1         19.0     19.0      0.0              run_results[validation_result_id] = run_result_obj
   343                                           
   344         2         29.0     14.5      0.0          return ValidationOperatorResult(
   345         1          1.0      1.0      0.0              run_id=run_id,
   346         1          1.0      1.0      0.0              run_results=run_results,
   347         1         10.0     10.0      0.0              validation_operator_config=self.validation_operator_config,
   348         1          1.0      1.0      0.0              evaluation_parameters=evaluation_parameters,
   349                                                   )
```

Note: I uploaded [my full lprof results](https://drive.google.com/file/d/1pWmdNAruZ3weEexwB9FPkewmI_nekldh/view?usp=sharing). It isn't very clear to me exactly what data is in these profile results, so I only shared this file with select individuals.

## vmprof Results

[vmprof](https://vmprof.readthedocs.io/en/latest/) is a **statistical** profiler and below results show **cpu times**, unlike the above two **deterministic profilers** showing **wall time**, so results are not directly comparable. Since this benchmark is currently I/O bound, these results may be misleading, but using vmprof may help if we decide to further optimize after we resolve the I/O bottleneck.

These profile results were from the baseline single threaded performance for a single table with the following command at https://github.com/great-expectations/great_expectations/pull/3280/commits/695cbfb71fed135e50e6966135111d06ccaf9925:

```
GE_VMPROF_FILE_PATH=tests/performance/results/vmprof_`date "+%H%M"`_`git rev-parse --verify HEAD`.prof pytest tests/performance/test_bigquery_benchmarks.py --no-spark --no-postgresql -rP -vv --bigquery --performance-tests -k "test_bikeshare_trips_benchmark[1-True]"
```

Some observations:
* 8.9% of CPU time is spent in [ActionListValidationOperator._run_actions](https://github.com/great-expectations/great_expectations/blob/fb20bd56e08b1aa0d7ef159c38b3dd822be1ea1b/great_expectations/validation_operators/validation_operators.py#L350).
* 6.8% of CPU time is spent in [DefaultJinjaView.render](https://github.com/great-expectations/great_expectations/blob/01b96541936583ac22dc13b34d22c9fb5802c1ac/great_expectations/render/view/view.py#L95).

call graph

Note: I uploaded [my full vmprof results](https://drive.google.com/file/d/1mzyUOK13eqYEix94MUBalk8sH2ZnutzO/view?usp=sharing). It isn't very clear to me exactly what data is in these profile results, so I only shared this file with select individuals.

### CODE REVIEW

**Comment by Shinnnyshinshin** [comment]
Hey @jdimatteo sorry about this. We are dealing with some dependency errors in our Azure pipelines, so I've cancelled the checks for now. Please feel free to try this again on Monday

**Comment by talagluck** [comment]
Thanks so much for this contribution, @jdimatteo ! We will review and discuss this internally as well, and come back with some feedback later this week.

**Comment by talagluck** [comment]
Thanks so much for this contribution, @jdimatteo ! We will review and discuss internally, and come back with some feedback later this week.

**Comment by cdkini** [comment]
Hey @jdimatteo! Just wanted to let you know that we're unfortunately dealing with some Azure issues (hence all the failures in the checks). Some dependency (or sub-dependency) changed on the 18th and is causing the pip resolver to really grind down to a halt.

Should have this cleaned up soon enough. Thanks for your patience and your continual contributions to GE

Edit: Things are looking good now!

File: great_expectations/checkpoint/checkpoint.py [hunk 1]
```diff
@@ -766,3 +735,101 @@ def run(
...
+                    _run_validation,
+                    f"{name}-checkpoint-validation[{idx}]",
+                    run_id,
+                    checkpoint_config,
+                    validation_dict,
+                    result_format,
+                    data_context,
+                )
+                validation_futures.append(val_op_run_future)
+
```
Comment (Shinnnyshinshin) [inline great_expectations/checkpoint/checkpoint.py:773]: Love this :) thank you

**Comment by jdimatteo** [comment]
The approach explored in this PR is implemented in a more robust way in https://github.com/great-expectations/great_expectations/pull/3362.

**Pull request closed without merging**